lib/core/src/Cardano/Pool/Metrics.hs

```
@@ -14,55 +14,155 @@
+    deriving (Show, Eq)
+
+-- | @worker@ follows the chain and puts pool productions and stake
+-- distributions to a @DBLayer@, such that the data in the @DB@ always is
+-- consistent.
+--
+-- The pool productions and stake distrubtions in the DB can /never/ be from
+-- different forks.
+worker
+    :: NetworkLayer IO t Block
+    -> DBLayer IO
+    -> Trace IO Text
+    -> IO ()
+worker nl db tr = do
+    -- Read the k latest headers the DB knows about
+    let (_block0, bp) = staticBlockchainParameters nl
+    let k = fromIntegral . getQuantity . view #getEpochStability $ bp
+    knownHeaders <- readCursor db k
+
+    follow nl tr knownHeaders advance' rollback getHeader
   where
-    slotInCurrentEpoch = ((epoch ==) . epochNumber)
+
+    advance'
+        :: NonEmpty Block
+        -> BlockHeader
+        -> IO (FollowAction ErrMetricsWorkerErr)
+    advance' bs h = handleResult <$> runExceptT (advance bs h)
+      where
+        handleResult = \case
+            Left ErrMetricsWorkerErrTipTooVolatile
+                -> Retry
+            Left e
+                -> ExitWith e
+                -- TODO: Make all cases explicit here.
+            Right ()
+                -> Continue
+
+    advance
+        :: NonEmpty Block
+        -> BlockHeader
+        -> ExceptT ErrMetricsWorkerErr IO ()
+    advance blocks nodeTip = do
+        -- Get a stake distribution and next batch of blocks such that we know
+        -- they are from the same chain.
+        --
+        -- Assumtion: We don't expect the node to switch from tip A to tip B,
+        -- and back to tip A.
+        --
+        -- If the tip is still the same after we have fetched the stake
+        -- distribution, we conclude that the stake-distrubtion is from the same
+        -- chain as the tip.
+        consistentDistr <- withinSameTip
+                (limitRetries 3)
+                (const $ throwE ErrMetricsWorkerErrTipTooVolatile)
+                getTip
+                $ \nodeTip2 -> do
```

suggestion:

```hs
let liftE = ...
consistentDist <- withinSameTip (limitRetries 3) liftE getTip $ ...
```

Also, it seems that we are simply swallowing any errors returned by the callback? `ErrMetricsWorkerStakeDistributionUnavailible` will never be returned.

---

Reverted to 
```haskell
          where
            liftE ErrWithinSameTipMaxRetries =
                throwE ErrMetricsWorkerErrTipTooVolatile

```

---

Isn't this a partial pattern match though...?